Sources/NIOPosix/SelectableEventLoop.swift

```diff
@@ -421,16 +430,169 @@ Further information:
+                        } else {
+                            nextScheduledTaskDeadline = task.readyTime
+                            moreScheduledTasksToConsider = false
+                        }
+                    } else {
+                        moreScheduledTasksToConsider = false
+                    }
+                }
+
+                assert(self.tasksCopy.count <= Self.taskCopyBatchSize)
+                // When we exit the loop, we would expect to
+                // * have taskCopy full, or:
+                // * to have completely drained task queues
+                //     * that means all _immediateTasks, and:
+                //     * all _scheduledTasks that are ready
+                assertExpression {
+                    if self.tasksCopy.count == Self.taskCopyBatchSize {
+                        return true
+                    }
+
+                    if !self._immediateTasks.isEmpty {
+                        return false
+                    }
+
+                    guard let nextScheduledTask = self._scheduledTasks.peek() else {
+                        return true
+                    }
+
+                    return nextScheduledTask.readyTime.readyIn(now) > .nanoseconds(0)
+                }
+
+                // If moreImmediateTasksToConsider is true, this means there must be more tasks
+                // in self._immediateTasks, or we've just filled up taskCopy on the last loop iteration
+                assertExpression {
+                    if !moreImmediateTasksToConsider {
+                        return true
+                    }
+
+                    if self.tasksCopy.count == Self.taskCopyBatchSize {
+                        return true
+                    }
+
+                    return self._immediateTasks.isEmpty
+                }
+
+                // Same as above, but for scheduled tasks
+                assertExpression {
+                    if !moreScheduledTasksToConsider {
+                        return true
+                    }
+
+                    if self.tasksCopy.count == Self.taskCopyBatchSize {
+                        return true
+                    }
+
+                    return self._scheduledTasks.isEmpty
+                }
+
+                // nextScheduledTaskDeadline should be set to > now, iff there are more
+                // scheduled tasks, and they are all scheduled for the future
+                // Moreover, nextScheduledTaskDeadline must equal the expiry time for the
+                // "top-most" scheduled task
+                assertExpression {
+                    if nextScheduledTaskDeadline == now {
+                        return true
+                    }
+
+                    guard let topMostScheduledTask = self._scheduledTasks.peek() else {
+                        return false
+                    }
+
+                    return topMostScheduledTask.readyTime == nextScheduledTaskDeadline
+                }
```

Comment:
Having all these assertions is great, but I feel like they're obscuring the algorithm itself, and the fact that they're inline closures gives them the ability to mutate the state (though they don't).

What about factoring these out to a `private static func` that runs all of these, and takes the state it validates as input (that is, does not pass `self`)? That would help make it really easy for future maintenance to see that these do not modify the state, and avoids these landing smack in the middle of the algorithm, but helps keep all their value.

Reply:
Actually, maybe an even more drastic refactor is available. A quick check of the current `main` suggests that `tasksCopy` being a stored property on `self` is entirely unnecessary: it's only ever accessed at `init` and in this method.

That suggests to me that we can refactor the entire `zip` logic into a `private static` helper function that takes the tasksCopy array `inout` and populates it, including testing its invariants. That would be a really nice refactor.

Reply:
Done, I introduced:
* [`_popTasksLocked`](https://github.com/apple/swift-nio/pull/2645/files#diff-565b354476ebe678da3be159a85e33e0b7630b1e45b62b20efe506210c1d62faR518): static function that takes all the state it needs to operate on `inout`
* [`_popTasksLockedAssertInvariants`](https://github.com/apple/swift-nio/pull/2645/files#diff-565b354476ebe678da3be159a85e33e0b7630b1e45b62b20efe506210c1d62faR444): static function that takes all the state after popping took place, and validates the invariants. The state is passed immutably.

Reply:
FYI @Lukasa, I made the  `tasksCopy` a local var in `runLoop`, but then made it a member variable back again in the [latest commit](https://github.com/apple/swift-nio/pull/2645/commits/c7a6eced58510463c2a1ce76288fed131227ba43). This is because otherwise allocation counter regresses (and shows up for tests that churn through lots of `SelectableEventLoop` instances).